development/webpack/utils/plugins/ManifestPlugin/index.ts

```diff
@@ -0,0 +1,348 @@
+import { extname, join } from 'node:path';
+import { readFileSync } from 'node:fs';
+import {
+  sources,
+  ProgressPlugin,
+  type Compilation,
+  type Compiler,
+  type Asset,
+} from 'webpack';
+import { validate } from 'schema-utils';
+import {
+  type DeflateOptions,
+  Zip,
+  AsyncZipDeflate,
+  ZipPassThrough,
+} from 'fflate';
+import { noop, type Manifest, Browser } from '../../helpers';
+import { schema } from './schema';
+import type { ManifestPluginOptions } from './types';
+
+const { RawSource, ConcatSource } = sources;
+
+type Assets = Compilation['assets'];
+
+const NAME = 'ManifestPlugin';
+const BROWSER_TEMPLATE_RE = /\[browser\]/gu;
+
+/**
+ * Clones a Buffer or Uint8Array and returns it
+ *
+ * @param data
+ * @returns
+ */
+function clone(data: Buffer | Uint8Array): Buffer {
+  return Buffer.from(data);
+}
+
+/**
+ * Adds the given asset to the zip file
+ *
+ * @param asset - The asset to add
+ * @param assetName - The name of the asset
+ * @param compress - Whether to compress the asset
+ * @param compressionOptions - The options to use for compression
+ * @param mtime - The modification time of the asset
+ * @param zip - The zip file to add the asset to
+ */
+function addAssetToZip(
+  asset: Buffer,
+  assetName: string,
+  compress: boolean,
+  compressionOptions: DeflateOptions | undefined,
+  mtime: number,
+  zip: Zip,
+): void {
+  const zipFile = compress
+    ? new AsyncZipDeflate(assetName, compressionOptions)
+    : new ZipPassThrough(assetName);
+  zipFile.mtime = mtime;
+  zip.add(zipFile);
+  // use a copy of the Buffer, as Zip will consume it
+  zipFile.push(asset, true);
+}
+
+/**
+ * A webpack plugin that generates extension manifests for browsers and organizes
+ * assets into browser-specific directories and optionally zips them.
+ *
+ * TODO: it'd be great if the logic to find entry points was also in this plugin
+ * instead of in helpers.ts. Moving that here would allow us to utilize the
+ * this.options.transform function to modify the manifest before collecting the
+ * entry points.
+ */
+export class ManifestPlugin<Z extends boolean> {
+  /**
+   * File types that can be compressed well using DEFLATE compression, used when
+   * zipping assets.
+   */
+  static compressibleFileTypes = new Set([
+    '.bmp',
+    '.cjs',
+    '.css',
+    '.csv',
+    '.eot',
+    '.html',
+    '.js',
+    '.json',
+    '.log',
+    '.map',
+    '.md',
+    '.mjs',
+    '.svg',
+    '.txt',
+    '.wasm',
+    '.vtt', // very slow to process?
```

Comment:
Kind of a funny comment to leave here. But I've left it here because over the months of working on this I kept bumping into files that were just slow, and I'd end up checking this list again and again.

It's just nice to know that it's known/expected. They are slow, but they are also compressible, so the comment is kind of comforting. :-)